### ISSUE TITLE
sddm : QtQuick is not installed

### ISSUE THREAD

**Issue by heydojo** [NixOS/nixpkgs#12132]
When I use ```services.xserver.displayManager.sddm.enable = true;```
sddm appears to start but a white screen appears and nothing more.
According to journalctl - module "QtQuick" is not installed. I see this message :

```sddm-greeter[896]: file:///nix/store/rdrx7fr26c2rz3hamkfw1f59skxdd552-sddm-0.13.0/share/sddm/themes/maui/Main.qml:26:1: module "QtQuick" is not installed 
                                              import QtQuick 2.0```

A quick google search for sddm white screen results in https://github.com/sddm/sddm/issues/262#issuecomment-54749222 and indicates a QtQuick version 2.2 import issue.

Other entries in journalctl which may or may not be relevant are :
*sddm-helper[896]: Could not open stderr to ".xsession-errors"
*sddm[635]: Greeter session started successfully
*sddm-greeter[896]: Socket error:  "QLocalSocket::connectToServer: Invalid name"
*sddm-greeter[896]: Cannot connect to the daemon - is it running?
*sddm-greeter[896]: file:///nix/store/rdrx7fr26c2rz3hamkfw1f59skxdd552-sddm-0.13.0/share/sddm/themes/maui/Main.qml:26:1: module "QtQuick" is not installed 
                                              import QtQuick 2.0
*sddm-greeter[896]: Adding view for "DFP9" QRect(0,0 1920x1080)

**Comment by heydojo** [comment]
Pinging @abbradar and @ttuegel in the hope that they notice this report.

**Comment by ttuegel** [comment]
Which version of NixOS are you using? Do you have KDE 5 enabled?

**Comment by heydojo** [comment]
cat /etc/os-release says :
```NAME=NixOS
ID=nixos
VERSION="16.03.git.c1be9baM (Emu)"
VERSION_ID="16.03.git.c1be9baM"
PRETTY_NAME="NixOS 16.03.git.c1be9baM (Emu)"
HOME_URL="http://nixos.org/"
```

I do not have KDE 5 enabled. My understanding is that sddm can be used with desktops other than KDE 5. I was hoping it was the case in NixOS. I was testing to see whether sddm would even start because it would not start for me due to issues I had with ati-drivers.

**Comment by abbradar** [comment]
Just my 2c: I'm using SSDM without KDE 5 on current `nixos-unstable` channel (`93d8671e2`).

**Comment by ttuegel** [comment]
> My understanding is that sddm can be used with desktops other than KDE 5. I was hoping it was the case in NixOS.

Yes, absolutely: SDDM and KDE are independent. However, KDE provides a theme for SDDM and if you enable KDE 5 and SDDM simultaneously, this theme will be used by default. I wanted to rule this out as a source of the problem before I investigated myself.

**Comment by heydojo** [comment]
Thanks for the replies. I really want to get onto KDE 5. :)

**Comment by ttuegel** [comment]
@heydojo I don't have revision `c1be9baM` in my Nixpkgs tree, and in fact, that doesn't seem to be a valid hash. What repository are you using?

**Comment by heydojo** [comment]
My fork of unstable from about a week ago which I branched and has an ati-driver update I am testing and no other modifications or merges.
https://github.com/heydojo/nixpkgs/tree/ati_unfree-crimson

**Comment by ttuegel** [comment]
This should be fixed in today's unstable channel update. Please ping me if the issue persists.

**Comment by heydojo** [comment]
Thanks for the update. I will try again asap.

**Comment by abbradar** [comment]
@ttuegel Can you please point out specific commits? I've run into the same problem on the new channel update.

**Comment by abbradar** [comment]
I think I've nailed it. For me https://github.com/abbradar/nixpkgs/commit/ae7c2c56c1ea1501a61309323ba8699666bb07eb fixed it -- @heydojo , can you please check if this patch is needed for you too?

**Comment by ttuegel** [comment]
@abbradar The wrapper should also pick up the dependency if you add `qtdeclarative` to the `propagatedBuildInputs` of the unwrapped package. I think there's a slight advantage to doing it that way in case anyone comes along later and wants to use `sddm.unwrapped` for some purpose.

Do you know how we can augment the NixOS test to detect this failure?

**Comment by abbradar** [comment]
I *think* we can grep logs, but I have practically no expertise in NixOS tests infrastructure.

**Comment by abbradar** [comment]
Also, would it be a good idea just to call `makeQtWrapper` inside `unwrapped` derivation?

**Comment by ttuegel** [comment]
Yeah, it used to do that, but then we can't have a KDE-themed version without rebuilding the whole thing.

**Comment by abbradar** [comment]
New version of the patch at https://github.com/abbradar/nixpkgs/commit/c41d4c11912ed7b5548a47b156f7f3e0082490e0

**Comment by heydojo** [comment]
Without the first patch posted by @abbradar I get a whitescreen. After applying the patch I can see the greeter and all looks normal but I type my password into the input box for my user account and press enter and nothing happens.
I found QIODevice::write(QLocalSocket): device not open in journalctl. So I now have a new issue which is preventing me from logging in. But so far I can actually see the greeter; which is a huge improvement.

**Comment by abbradar** [comment]
Patch pushed to the `master`, thanks. Meanwhile I cannot reproduce `QIODevice` problem at all -- login works nicely for me.

If that helps: I have `services.xserver.desktopmanager.xfce` enabled. Also I have a strange race condition which haunts me to this day -- sometimes `sddm` gives me a black screen with a cursor, workarounded by setting `autorun = false` and starting it by myself. _Maybe_ this also affects you -- try setting `autorun = false` and doing `systemctl start display-manager`.

Sadly, I've found no clues in logs about this problem -- if somebody has any ideas, please share.

**Comment by heydojo** [comment]
```
Sadly, I've found no clues in logs about this problem -- if somebody has any ideas, please share.
```
Strongly suspect systemd.

**Comment by abbradar** [comment]
Me too, I've seen something `systemd-logind`-related in logs but I have no idea what could it mean, and the bug hits rarely and usually at a very inconvenient time to debug it.

**Comment by heydojo** [comment]
https://wiki.archlinux.org/index.php/improve_boot_performance
A systemd boot chart is a nice way to see the duration and order of each systemd unit.
Then using that information you can "tune" units so that they start or do not start before or after another unit.

I guess because systemd is self optimising; that sometimes it starts the display manager unit a fraction too early because it has not been told to wait for the xserver. Or something like that.
It reminds me of why the developers want kdus in the kernel.

**Comment by abbradar** [comment]
Aha, I haven't thought about using bootchart for debugging race conditions -- nice idea!

Guess I'll make some script to record bootcharts for last N boots, wait until the problem hits again and look at the results.

**Comment by abbradar** [comment]
Well, shall we close this issue as solved and move to another one regarding your new problem?

**Comment by ttuegel** [comment]
> Also I have a strange race condition which haunts me to this day -- sometimes sddm gives me a black screen with a cursor, workarounded by setting autorun = false and starting it by myself.

I haven't seen this, but I usually run the KDE theme, which takes longer to load.

One thing that jumps out at me is that `display-manager.service` is *not* instructed to wait for `systemd-logind.service`. Even if that is not the cause of your issue, it is probably not correct.

**Comment by abbradar** [comment]
Bullseye! I've tried rebooting the PC five times with and without the patch -- 4/5 fails without, 0/5 with.

**Comment by heydojo** [comment]
@ttuegel which qt module provides Qt Network? Because that's the dependency which I think needs to be added to the build script for sddm to let me login.

**Comment by ttuegel** [comment]
@heydojo `qtbase` provides Qt Network. We could try adding it to `propagatedBuildInputs` [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/display-managers/sddm/default.nix#L31), but it's already in `buildInputs` and I don't think it provides any *runtime* dependencies.

**Comment by heydojo** [comment]
Yeah, it's not qtbase.
Are there any systemd sockets in play between sddm, pam, systemd and systemd-logind which might be causing starting too early or late?

**Comment by ttuegel** [comment]
@heydojo Well, `systemd-logind` definitely needs to be started before `sddm`. Are you using @abbradar's latest patches for it from the `master` branch? It would also help to know if you are just logging in as a regular local user, or through a network account (LDAP, Kerberos, etc.).

**Comment by heydojo** [comment]
I have not pulled today so no I have not yet tested the patch. I am logging in as a local user.
According to https://github.com/sddm/sddm/wiki/Installation I am guessing I should try and propagate
libxcb and or pam?
I will pull the patch @abbradar committed and try again.

**Comment by abbradar** [comment]
Wild guess: `rm -rf /var/lib/sddm`. Somebody here a long time ago had some problem that was resolved like this ~_^.

**Comment by ttuegel** [comment]
> I am guessing I should try and propagate libxcb and or pam?

I would try @abbradar's suggestion first. What you suggest is unlikely to help. Propagating dependencies just makes them available at runtime; however, libxcb doesn't provide any runtime features as far as I know, and pam is already installed system-wide.

**Comment by heydojo** [comment]
Yeah I tried the latest from unstable, deleting /var/lib/sddm and still I cannot login.
I tried using the open source radeon driver and still no luck.

Is it a reasonable guess that it could be an authentication problem related to the systemd units which start sddm and the desktop-manager? 

Is it possible that the sddm systemd unit needs :
```
security.pam.services.kde = { allowNullPassword = true; startSession = true; };
```
adding to it's security.pam.services entry to be able to send the password?

**Comment by heydojo** [comment]
I cannot login if I select xfce session either.

I am going to close this because the issue is fixed. And I will open a new report for the new issue.

**Comment by wedens** [comment]
I have the same problem with `QtQuick` in recent unstable (d0e3cca).

**Comment by abbradar** [comment]
You need a patch which was commited a bit later -- https://github.com/NixOS/nixpkgs/commit/5ad58df002dbb0e35b5faac0b153a8413998c7e2.
